File: planik-android/app/src/main/java/pl/planik/domain/source/PlanLocalSource.kt
package pl.planik.domain.source

import androidx.paging.PagingData
import kotlinx.coroutines.flow.Flow
import pl.planik.domain.model.DayEntry
import pl.planik.domain.model.NewDayEntry
import pl.planik.domain.model.NewPlan
import pl.planik.domain.model.Plan
import pl.planik.domain.model.UpdateDayEntry

interface PlanLocalSource {
  suspend fun hasPlan(): Boolean
  fun observeCurrentPlan(userId: Int): Flow<Plan?>
  fun observePlan(id: Int, currentUserId: Int): Flow<Plan?>
  suspend fun getPlan(id: Int, userId: Int): Plan?
  fun pagedPlans(pageSize: Int): Flow<PagingData<Plan>>
  suspend fun createPlan(userId: Int, newPlan: NewPlan): Int
  suspend fun updatePlan(plan: Plan): Int?
  suspend fun deletePlan(id: Int)
  suspend fun getDayEntry(dayEntryId: Int, planId: Int): DayEntry?
  suspend fun addDayEntry(planId: Int, newDayEntry: NewDayEntry): Int
  suspend fun updateDayEntry(
    dayEntryId: Int,
    planId: Int,
    dayEntry: UpdateDayEntry
  ): Int?
}

File: planik-flutter/planik/lib/domain/source/plan_local_source.dart
import 'package:fpdart/fpdart.dart';
import 'package:planik/domain/failure/general_failure.dart';
import 'package:planik/domain/model/full_plan.dart';
import 'package:planik/domain/model/plan.dart';

abstract class PlanLocalSource {
  TaskEither<GeneralFailure, FullPlan> getPlan({required int id});

  TaskEither<GeneralFailure, List<Plan>> getPlans({required int pageSize, required int page});

  TaskEither<GeneralFailure, int> createPlan({
    required int userId,
    required String name,
    required bool current,
  });
}
